pkg/logql/shardmapper.go

```diff
@@ -0,0 +1,224 @@
+package logql
+
+import (
+	"fmt"
+
+	"github.com/cortexproject/cortex/pkg/querier/astmapper"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+)
+
+func NewShardMapper(shards int) (ShardMapper, error) {
+	if shards < 2 {
+		return ShardMapper{}, fmt.Errorf("Cannot create ShardMapper with <2 shards. Received %d", shards)
+	}
+	return ShardMapper{shards}, nil
+}
+
+type ShardMapper struct {
+	shards int
+}
+
+func (m ShardMapper) Map(expr Expr) (Expr, error) {
+	switch e := expr.(type) {
+	case *literalExpr:
+		return e, nil
+	case *matchersExpr, *filterExpr:
```

Comment:
I'd argue `marchersExpr` requires shard mapping. Those will just increase the amount of data we process with their overhead, this is because the result is limited and it will always be super fast to answer unless there is a lot streams in the query.

However here it seems that the function is recursive so in case of metrics `rate({app="foo"}[5m])` we probably wants sharding mapping.

It's also difficult to imagine because I have yet to discover where will the parallelization take place.  So I'll need more to judge on that.

Reply:
Good point. I'll think about how to more effectively short circuit this rather than relying on the frontend.